apollo-backend/src/main/java/io/logz/apollo/controllers/EnvironmentsStackController.java

```diff
@@ -0,0 +1,118 @@
+    public EnvironmentsStack getEnvironmentsStack(int stackId) {
+        EnvironmentsStack environmentsStack = stackDao.getEnvironmentsStack(stackId);
+        environmentsStack.setEnvironments(environmentsStackDao.getEnvironments(stackId));
+        return environmentsStack;
+    }
+
+    @LoggedIn
+    @GET("/environments-stack")
+    public List<EnvironmentsStack> getAllEnvironmentsStacks() {
+        List<EnvironmentsStack> environmentStacks = stackDao.getAllEnvironmentStacks();
+        return environmentStacks.stream().map(stack -> {
+            List<Integer> environments = environmentsStackDao.getEnvironments(stack.getId());
+            return new EnvironmentsStack(stack.getId(), stack.getName(), stack.isEnabled(),environments);
+        }).collect(Collectors.toList());
+    }
+
+    @LoggedIn
+    @POST("/environments-in-stacks")
+    public void addEnvironmentToStack(int environmentId, int stackId, Req req) {
+        environmentsStackDao.addEnvironmentToStack(environmentId, stackId);
+        assignJsonResponseToReq(req, HttpStatus.CREATED, stackId);
+    }
+
+    @LoggedIn
+    @Transactional
+    @POST("/environments-stack")
+    public void addEnvironmentsStack(String name, String isEnabled, String environmentsCsv, Req req) {
+        Iterable<String> environmentsString = Splitter.on(ENVIRONMENTS_DELIMITER).omitEmptyStrings().trimResults().split(environmentsCsv);
+        List<Integer> environmentsIds = StreamSupport.stream(environmentsString.spliterator(), false)
+                                                     .map(environment -> Integer.parseInt(environment))
+                                                     .collect(Collectors.toList());
+        EnvironmentsStack environmentsStack = new EnvironmentsStack(name, Boolean.valueOf(isEnabled));
+        stackDao.addStack(environmentsStack);
+        if (environmentsIds.size() > 0) {
+            environmentsStackDao.addEnvironmentsToStack(environmentsIds, environmentsStack.getId());
+        }
+        assignJsonResponseToReq(req, HttpStatus.CREATED, environmentsStack.getId());
+    }
+
+    @LoggedIn
+    @Transactional
+    @PUT("/environments-stack")
+    public void updateEnvironmentsStack(int id, String name, boolean isEnabled, String environmentsCsv, Req req) {
+        Iterable<String> environmentsString = Splitter.on(ENVIRONMENTS_DELIMITER).omitEmptyStrings().trimResults().split(environmentsCsv);
+        List<Integer> environmentsIds = StreamSupport.stream(environmentsString.spliterator(), false)
+                                                     .map(environment -> Integer.parseInt(environment))
+                                                     .collect(Collectors.toList());
+        EnvironmentsStack environmentsStack = new EnvironmentsStack(id, name, isEnabled);
+        stackDao.updateStack(environmentsStack);
+        environmentsStackDao.clearEnvironmentsStack(id);//TODO check with mesika
+        if (environmentsIds.size() <= 0) {
```

Comment:
Do that before you start the update.. that way it can be in a limbo state.
If the request is bad, its expected that nothing else will happen before